man/man1/lgt2svg.1
.TH lgt2svg 1 "November 19, 2024" "Logtalk 3.86.0" "Logtalk Documentation"

.SH NAME
lgt2svg \- Logtalk script for converting .d2 and .dot diagram files into SVG files

.SH SYNOPSIS
.B lgt2svg
[\fB-c \fIcommand\fR]
[\fB-- \fIarguments\fR]
.br
.B lgt2svg
[\fB-l \fIlayout\fR]
[\fB-- \fIarguments\fR]
.br
.B lgt2svg
.B \-v
.br
.B lgt2svg
.B \-h

.SH DESCRIPTION
\fBlgt2svg\fR is a Logtalk script for converting .d2 and .dot diagram files generated by the \fBdiagrams\fR tool in the current directory into SVG files.

.SH OPTIONS
.TP
.B \-c
Graphviz command (valid values are \fIdot\fR, \fIcirco\fR, \fIfdp\fR and \fIneato\fR; default is \fIdot\fR).
.TP
.B \-l
d2 layout (valid values are \fIdagre\fR, \fIelk\fR and \fItala\fR; default is \fIelk\fR).
.TP
.BI \--
Arguments to be passed to the converter command (no default).

.SH REQUIREMENTS
Requires recent versions of Graphviz and d2 installed with the command-line executables available from the system \fBPATH\fR.

.SH ENVIRONMENT
.TP
.B LOGTALKHOME
Specifies the location of the Logtalk installation.
.TP
.B LOGTALKUSER
Specifies the location of the Logtalk user directory.

.SH FILES
.TP
.BI $LOGTALKUSER/tools/diagrams/diagrams.css
CSS file styling the SVG diagram files.

.SH "SEE ALSO"
The Logtalk website at https://logtalk.org/
.PP
The Logtalk diagrams tool documentation at https://logtalk.org/manuals/devtools/diagrams.html

.SH AUTHOR
Paulo Moura <pmoura@logtalk.org>

.SH COPYRIGHT
Copyright (c) 1998-2024 Paulo Moura.


RELEASE_NOTES.md
________________________________________________________________________

This file is part of Logtalk <https://logtalk.org/>  
SPDX-FileCopyrightText: 1998-2024 Paulo Moura <pmoura@logtalk.org>  
SPDX-License-Identifier: Apache-2.0

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
________________________________________________________________________


RELEASE NOTES
=============


3.86.0 - December ??, 2024
==========================

Prolog adapter and integration files
------------------------------------

* UPDATED: Ciao Prolog adapter file to pretend that modules are supported
to allow loading libraries and examples that require file `use_module/1`
directives.

Documentation
-------------

* ADDED: Notes on encrypting applications using XVM.

Tools
-----

* CHANGED: The `diagrams` tool now uses a Unicode glyph for the magnifying
lens instead of a PNG image.

* REMOVED: Deprecated `lgt2svg.js` script from the `diagrams` tool.

* ADDED: Experimental `diagrams` tool support for the d2 graph language.

* UPDATED: The `diagrams` tool to support graph languages such as d2 where the
node identifiers are not global.

* UPDATED: The `diagrams` tool to also support PearAI editor local file links.

* FIXED: The `diagrams` tool to also generate an `instance/class` caption for
metaclasses.

* FIXED: Bug in the `diagrams` tool where repeated nodes could be created for
external entities.

Examples
--------

* UPDATED: The `assumptions` example with another individual example.


3.85.0 - November 12, 2024
==========================

Logtalk compiler and runtime
----------------------------

* CHANGED: The `logtalk_load_context/2` built-in predicate to type-check
the key argument.

* CHANGED: Attempting to define an entity/predicate alias equal to the
original entity/predicate is now a compilation error instead of simply
a warning.

* CHANGED: Reclassified the `underscore_variables` flag as a backend flag
instead of a linter flag.

* ADDED: New meta-flag, `linter`, to manage the values of all the linter
flags as a group.

* ADDED: New linter flag, `encodings`, to control source file text encoding
warnings.

* ADDED: New linter flag, `general`, for warnings that are not controlled
by a specific flag.

* IMPROVED: Print a linter warning on redundant entity qualification in
predicate directives.

* IMPROVED: Detect and report as compilation errors `use_module/1-2` and
`uses/1-2` directives when compiling a protocol.

* IMPROVED: Report as compilation errors instead of compilation warnings
`use_module/1-2` directives when not compiling a module as an object while
using a backend that doesn't support modules.

* IMPROVED: Report as compilation errors `export/1` and `reexport/2`
directives when not compiling a module as an object.

* FIXED: The compiler to prevent trying to compile a Prolog `use_module/1`
directive as a Logtalk directive when compiling a module as an object.

* FIXED: Detection of a file trying to recursively load itself to not throw
a permission error when reloading a file that previously failed to load.

* FIXED: Misleading error when compiling `dynamic/1`, `discontiguous/1`, and
`multifile/1` directives with arguments with redundant entity qualification.

Prolog adapter and integration files
------------------------------------

* UPDATED: Trealla Prolog adapter file to require version 2.59.21 or later.

Documentation
-------------

* IMPROVED: Documentation of the `issue_creator` tool.

* IMPROVED: The Handbook documentation on performance, adding a section on
circular references.

* IMPROVED: The Handbook documentation on the `logtalk_make/0-1` built-in
predicates.

* UPDATED: The Handbook documentation on the `underscore_variables` flag.

Library
-------

* ADDED: Additional test for the `java` library checking setting an array
element to the empty list atom.

* UPDATED: The `meta_compiler` library to support optimization of meta-calls
with closure meta-arguments that are messages to parametric objects.

Tools
-----

* ADDED: Support to the `diagrams` tool for VSCode, VSCodium, Cursor, Zed,
BBEdit, MacVim, and TextMate local file links.

* UPDATED: Rewrite error and warning messages from the `packs` tool so that
they can be explained by the `tutor` tool.

* UPDATED: The `tutor` tool to explain the new redundant entity qualifier 
linter warning.

* FIXED: Case where the `packs` tool would print repeated "Pack dependency
not available" error messages when trying to install or update a pack with
alternative dependencies where none is available.

* FIXED: The `debugger` tool description of the `p` (print) port command.

* FIXED: Typos in the `mode/2` directives for the `debuggerp` protocol
`logging/3` and `nolog/3` public predicates.

* FIXED: Typo in the `info/2` directive for the `tutor` object `explain//1`
public non-terminal.

* FIXED: Harmless linter warning when loading the `debugger` tool with some
backends.

Examples
--------

* IMPROVED: The `searching` example representation of some of the problems.

* UPDATED: The `errors` example for the new redundant entity qualifier linter
warning.

* FIXED: Portability issue in the `errors` example tests.

* FIXED: Cannot run the `module_aliases` example with XSB and Tau Prolog.

* FIXED: The `symbiosis` example for Trealla Prolog removing its `apply`
module.

Tests
-----

* CHANGED: Tests for the Prolog standard `initialization/1` directive to not
depend on support for the `ensure_loaded/1` directive.

* ADDED: Additional tests for the `user` built-in entity.


tools/diagrams/lgt2svg.sh
#!/usr/bin/env bash

#############################################################################
## 
##   DOT diagram files to SVG files conversion script 
##   Last updated on November 19, 2024
## 
##   This file is part of Logtalk <https://logtalk.org/>  
##   SPDX-FileCopyrightText: 1998-2024 Paulo Moura <pmoura@logtalk.org>
##   SPDX-License-Identifier: Apache-2.0
##   
##   Licensed under the Apache License, Version 2.0 (the "License");
##   you may not use this file except in compliance with the License.
##   You may obtain a copy of the License at
##   
##       http://www.apache.org/licenses/LICENSE-2.0
##   
##   Unless required by applicable law or agreed to in writing, software
##   distributed under the License is distributed on an "AS IS" BASIS,
##   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
##   See the License for the specific language governing permissions and
##   limitations under the License.
## 
#############################################################################


if ! [ "$LOGTALKHOME" ]; then
	echo "The environment variable LOGTALKHOME should be defined first, pointing"
	echo "to your Logtalk installation directory!"
	echo "Trying the default locations for the Logtalk installation..."
	if [ -d "/usr/local/share/logtalk" ]; then
		LOGTALKHOME=/usr/local/share/logtalk
		echo "... using Logtalk installation found at /usr/local/share/logtalk"
	elif [ -d "/usr/share/logtalk" ]; then
		LOGTALKHOME=/usr/share/logtalk
		echo "... using Logtalk installation found at /usr/share/logtalk"
	elif [ -d "/opt/local/share/logtalk" ]; then
		LOGTALKHOME=/opt/local/share/logtalk
		echo "... using Logtalk installation found at /opt/local/share/logtalk"
	elif [ -d "/opt/share/logtalk" ]; then
		LOGTALKHOME=/opt/share/logtalk
		echo "... using Logtalk installation found at /opt/share/logtalk"
	elif [ -d "$HOME/share/logtalk" ]; then
		LOGTALKHOME="$HOME/share/logtalk"
		echo "... using Logtalk installation found at $HOME/share/logtalk"
	elif [ -f "$( cd "$( dirname "$0" )" && pwd )/../core/core.pl" ]; then
		LOGTALKHOME="$( cd "$( dirname "$0" )" && pwd )/.."
		echo "... using Logtalk installation found at $( cd "$( dirname "$0" )" && pwd )/.."
	else
		echo "... unable to locate Logtalk installation directory!" >&2
		echo
		exit 1
	fi
	echo
elif ! [ -d "$LOGTALKHOME" ]; then
	echo "The environment variable LOGTALKHOME points to a non-existing directory!" >&2
	echo "Its current value is: $LOGTALKHOME" >&2
	echo "The variable must be set to your Logtalk installation directory!" >&2
	echo
	exit 1
fi
export LOGTALKHOME

if ! [ "$LOGTALKUSER" ]; then
	echo "The environment variable LOGTALKUSER should be defined first, pointing"
	echo "to your Logtalk user directory!"
	echo "Trying the default location for the Logtalk user directory..."
	export LOGTALKUSER=$HOME/logtalk
	if [ -d "$LOGTALKUSER" ]; then
		echo "... using Logtalk user directory found at $LOGTALKUSER"
	else
		echo "... Logtalk user directory not found at default location. Creating a new"
		echo "Logtalk user directory by running the \"logtalk_user_setup\" shell script:"
		logtalk_user_setup
	fi
elif ! [ -d "$LOGTALKUSER" ]; then
	echo "Cannot find \$LOGTALKUSER directory! Creating a new Logtalk user directory"
	echo "by running the \"logtalk_user_setup\" shell script:"
	logtalk_user_setup
fi
echo


print_version() {
	echo "$(basename "$0") 0.12"
	exit 0
}


# default argument values
command="dot"
layout="elk"


usage_help()
{
	echo 
	echo "This script converts .d2 and .dot files in the current directory to SVG files"
	echo
	echo "Usage:"
	echo "  $(basename "$0") [-c command] [-- arguments]"
	echo "  $(basename "$0") [-l layout] [-- arguments]"
	echo "  $(basename "$0") -v"
	echo "  $(basename "$0") -h"
	echo
	echo "Optional arguments:"
	echo "  -c Graphviz command (dot, circo, fdp, or neato; default is $command)"
	echo "  -l d2 layout (dagre, elk, or tala; default is $layout)"
	echo "  -- additional arguments to be passed to the converter command (no default)"
	echo "  -v print version"
	echo "  -h print help"
	echo
}


while getopts "c:l:vh" Option
do
	case $Option in
		c) c_arg="$OPTARG";;
		l) l_arg="$OPTARG";;
		v) print_version;;
		h) usage_help; exit;;
		*) usage_help; exit;;
	esac
done

shift $((OPTIND - 1))
args=("$@")


if [ "$c_arg" == "dot" ] ; then
	command="dot"
elif [ "$c_arg" == "circo" ] ; then
	command="circo"
elif [ "$c_arg" == "fdp" ] ; then
	command="fdp"
elif [ "$c_arg" == "neato" ] ; then
	command="neato"
elif [ "$c_arg" != "" ] ; then
	echo "Error! Unknown Graphviz command: $c_arg" >&2
	usage_help
	exit 1
fi

if [ "$l_arg" == "dagre" ] ; then
	layout="dagre"
elif [ "$l_arg" == "elk" ] ; then
	layout="elk"
elif [ "$l_arg" == "tala" ] ; then
	layout="tala"
elif [ "$l_arg" != "" ] ; then
	echo "Error! Unknown d2 layout: $l_arg" >&2
	usage_help
	exit 1
fi

d2_count=$(ls -1 ./*.d2 2>/dev/null | wc -l)
dot_count=$(ls -1 ./*.dot 2>/dev/null | wc -l)

d2_failed_flag=0
dot_failed_flag=0

if [ $d2_count -ne 0 ] ; then
	echo "Converting .d2 files to .svg files ..."
	cp "$LOGTALKUSER/tools/diagrams/diagrams.css" .
	for file in ./*.d2; do
		echo -n "  converting $(basename "$file")... "
		d2 --layout "$layout" "${args[@]}" "$file" "${file%.*}.svg"  2>/dev/null | cat
		if [ "${PIPESTATUS[0]}" == 0 ] ; then
			echo "done"
		else
			d2_failed_flag=1
			echo "failed"
		fi
	done
fi

if [ $dot_count -ne 0 ] ; then
	echo "Converting .dot files to .svg files ..."
	cp "$LOGTALKUSER/tools/diagrams/diagrams.css" .
	for file in ./*.dot; do
		echo -n "  converting $(basename "$file") "
		converted=1
		counter=24
		while [ $converted -eq 1 ] && [ $counter -gt 0 ] ; do
			$command -q -Tsvg -Gfontnames=svg -o "${file%.*}.svg" "${args[@]}" "$file" 2>/dev/null | cat
			if [ "${PIPESTATUS[0]}" == 0 ] ; then
				converted=0
			fi
			(( --counter ))
			echo -n "."
		done
		if [ $counter == 0 ] ; then
			dot_failed_flag=1
			echo " failed"
		else
			echo " done"
		fi
	done
fi

if [ $d2_count -eq 0 ] && [ $dot_count -eq 0 ] ; then
	echo "No .d2 or .dot files exist in the current directory!"
	echo
fi

if [ $d2_failed_flag -eq 0 ] && [ $dot_failed_flag -eq 0 ] ; then
	echo "Conversion done."
	echo
	exit 0
else
	echo "One or more files could not be converted!"
	echo
	exit 1
fi
